### ISSUE TITLE
Modifiers are not applied when loading scenes.

### ISSUE THREAD

**Issue by blenderaiser** [armory3d/armory#1466]
**Description**
When I load a scene through 'set' or 'spawn', the modifiers on its objects are not applied. I tested through nodes as well as scripts, and the same result happened.
Interestingly, it works sometimes, however I need to runs separately each scene, and then run the main scene.
*This modifiers problem always occurs in the exported game.

**To Reproduce**
- Create a blend file.
- Create a new scene.
- Create a cube with the bevel modifier in the new scene.
- Return to the main scene.
- Create an empty object and attach a node tree like this:
![Screenshot_2019-11-05_18-49-58](https://user-images.githubusercontent.com/13911941/68256431-e03f2c00-000e-11ea-8574-ae20b42b79ec.png)
- Run F5 and press space.

**Expected behavior**
Modifiers should be applied to the loaded scene even when the scene is in another blend file (via link).

**System**
Blender: 2.8
Armory: 2019-11
OS: Windows 7 and Linux (Manjaro)
Graphics card: geforce gtx 750 ti

**Test File**
[test.zip](https://github.com/armory3d/armory/files/3811853/test.zip)

**Comment by N8n5h** [comment]
I could reproduce it on my end thanks to your file :+1:

I was digging a bit on the Scene class to see if I could spot where the issue may be, but AFAIK Armory "apply" modifiers (by obtaining an evaluated copy from the dependency graph) when exporting meshes here: https://github.com/armory3d/armory/blob/3e57614e1254cd25f207be9c6d4bf8e5f211b0bc/blender/arm/exporter.py#L1269
So meshes should be at their "exported state" when loading scenes, regardless of being on the main scene or not.

My bet is that the dependency graph is not giving a correct copy of the object with "applied" modifiers for different scenes given the current trend: https://github.com/armory3d/armory/issues/1378#issuecomment-531398321

**Comment by blenderaiser** [comment]
Thanks for your attention @N8n5h!

I made a humble example of a possible solution :P 
Open the main.blend and just run the script and look at the terminal to see the count of vertices with the modifiers applied.
Do you think this example will be helpful?

*If I'm not mistaken, the Unity3D also open each scene in exportation. But as there are few objects in the example, this is not even perceptible.
[Blender - Eval All Scenes.zip](https://github.com/armory3d/armory/files/3822470/Blender.-.Eval.All.Scenes.zip)

**Comment by N8n5h** [comment]
So it seems I was right on my guess. Your solution could fit as a workaround for now :+1:  but ideally the bigger issue with the dependency graph should be adressed at some moment; maybe it could be reported to Blender directly if it's still there in the latest versions, the only complicated thing is to isolate this issue (to reproduce it without Armory basically) and prove that Blender is the issue and not Armory. 

If you don't have a problem I will try to apply your solution to the exporter and send a PR so it works for the moment, is that ok?

**Comment by blenderaiser** [comment]
I tested the branch with my example and it worked! =]

I also tested on another example that worked, but I had to change some modifier parameters and save the file ... weird. It may be some mistake on my part too.

I will use this solution in my projects and analyze the consistency of it.

**Comment by N8n5h** [comment]
> I also tested on another example that worked, but I had to change some modifier parameters and save the file ... weird. It may be some mistake on my part too.

It seems like maybe the fix wasn't enough to workaround the depency graph issue...

**Comment by DevOnmai** [comment]
The solution is working well for scenes in the same file and linked objects, but linked scenes are not.

Example to run with the branch:
[test_solution.zip](https://github.com/armory3d/armory/files/3826247/test_solution.zip)

**Comment by blenderaiser** [comment]
The solution is working well for scenes in the same file and linked objects, but linked scenes are not.

Example to run with the branch:
[test_solution.zip](https://github.com/armory3d/armory/files/3826270/test_solution.zip)

**Comment by N8n5h** [comment]
I tested your files and I'm not sure why the solution doesn't work... I think is something it's ought to be reported to the Blender devs, no point in working around it.

**Comment by blenderaiser** [comment]
I understand, but the test code mentioned below works correctly in all scenarios (I did more tests). Wouldn't the problem be at Armory?

> Thanks for your attention @N8n5h!
> 
> I made a humble example of a possible solution :P
> Open the main.blend and just run the script and look at the terminal to see the count of vertices with the modifiers applied.
> Do you think this example will be helpful?
> 
> *If I'm not mistaken, the Unity3D also open each scene in exportation. But as there are few objects in the example, this is not even perceptible.
> [Blender - Eval All Scenes.zip](https://github.com/armory3d/armory/files/3822470/Blender.-.Eval.All.Scenes.zip)

**Comment by N8n5h** [comment]
>  but the test code mentioned below works correctly in all scenarios

Well the only thing I did was change the scene before doing any exporting to apply something similar to your solution, but I could have missed something, you are free to check where the process might fail in the exporter: https://github.com/N8n5h/armory/blob/patch-1/blender/arm/make.py. 

> Wouldn't the problem be at Armory?

I personally don't have a lot of time in hand right now to do tests to spot where the issue lies in the exporter, but I'm inclined to believe the problem is right around the dependency graph since I manually tested it before some time ago and found that it caused problems in another issue, but I could be wrong this time.

**Comment by Sandy10000** [comment]
It seems that additional setting is necessary when data of another file is linked and used.(https://github.com/armory3d/armory/wiki/multiuser)

**Comment by N8n5h** [comment]
I tried to test what Sandy10000 posted, but I'm having issues with the text and the solidify modifier, so I'm unsure if text is an ideal subject for testing this, I feel like with an simpler object I could get more consistent results.

@blenderaiser  Can you recreate the tests in `Blender - Eval All Scenes.zip` but with a mesh object instead? I tried to recreate it myself but failed because I'm not sure how you linked the objects in the same scene so that they appear to be with a "lock" in the UI.

**Comment by blenderaiser** [comment]
The links are made like this: File -> Link -> Blend File -> Scene -> Scene
... or for objects: File -> Link -> Blend File -> Object -> Cube
That's why they get blocked.

I could use proxies, but they don't work well for objects with children.
So for more than one person to edit the same project and have something similar to prefabs I intend to use a lot of links.

I put some logs in make.py and exporter.py and I found a problem.  Armory is evaluating linked scene objects when the main scene is defined in bpy.context.window.scene. In the print you can see it. I marked in red the objects that were evaluated in the wrong scene, they should be evaluated in 'Linked Scene'.
![Screenshot_2019-11-10_12-44-56](https://user-images.githubusercontent.com/13911941/68548432-9ae27c00-03cb-11ea-9614-3547b15d4a8e.png)

Logs with '------- My Code' show 100% correctly. This code I put in the export_objects function in exporter.py:
```python 
...
self.output['mesh_datas'] = []

for object in scene.objects:
       if object.type == "MESH":
           evobj = object.evaluated_get(self.depsgraph)
           mesh = evobj.to_mesh()
           print("----- My Code: " + evobj.name + " | original vertices: " + str(len(object.data.vertices)) +", evaluated vertices: "+str(len(mesh.vertices)))
           evobj.to_mesh_clear()
```

The log with '******* Armory Code' in put in export_mesh function in exporter.py:
```python 
...
exportMesh = bobject_eval.to_mesh()
print("***** Armory Code: " + bobject.name + " | original vertices: " + str(len(bobject.data.vertices)) + ", evaluated vertices: " + str(len(exportMesh.vertices)))
```

But I still don't know why these errors occur.
Hope this helps in something.

**Comment by blenderaiser** [comment]
@N8n5h, running F5, using the codes from the previous post, I believe you will see the same behavior in the terminal.
[Blender - Eval All Scenes 2.zip](https://github.com/armory3d/armory/files/3828862/Blender.-.Eval.All.Scenes.2.zip)

**Comment by N8n5h** [comment]
Yeah, it seems like my quick fix didn't do much. I was reading the code and it's a bit more complicated than I thought; it would require me to sit and do a lot more of reading to understand how the exporter works.

**Comment by N8n5h** [comment]
Ok, digging again in the exporter source I found this https://github.com/armory3d/armory/blob/3e57614e1254cd25f207be9c6d4bf8e5f211b0bc/blender/arm/exporter.py#L1944
After commenting those lines the linked objects are exported in their correct context I presume, same with the dependency graph.
I tested it with the files `Blender - Eval All Scenes 2.zip` but still doesn't works :disappointed:

**Comment by blenderaiser** [comment]
> I discovered the issue with my quick fix, it seems that bpy.context.scene is not being updated when changing the scene with bpy.context.window.scene = scene and instead it stays always in the same scene for some reason.

I reached the same conclusion! =]

I am creating an exporter for Panda3d, for study only, and managed to solve this problem apparently. But I still don't know how to implement this in Armory. 

The secret was to use the bpy.types.WindowManager.event_timer_add function on the export operator of all scenes. In a timer step you load the scene and next you evaluate the objects....

You can also run evaluated_depsgraph_get in the timer step until the scene in depsgraph.scene_eval is the desired scene, and only then evaluate the objects.

I have no time now, but I plan to make an example soon and post here.

**Comment by N8n5h** [comment]
Update: I found where the issue might be. It seems that when calling an operator, in this case `arm.play` (ArmoryPlayButton), from a panel, in this case Player panel (ARM_PT_ArmoryPlayerPanel), it has some effect in `bpy.context.scene` that prevents it to work correctly.

If you call the operator directly from the search window (space in Blender > search for "Play" > click "arm: Play"), `bpy.context.scene` will work just fine after the scene change.

![image](https://user-images.githubusercontent.com/42382648/68954084-0a78b280-07a2-11ea-8dbb-7d0d4536fe4e.png)

I isolated the issue in a small blend file and reported it to the blender devs here: https://developer.blender.org/T71542

If I have some time I will try to dig Blender's source to see if I can spot the issue there...

**Comment by N8n5h** [comment]
Update: It seems that the problem is not the panel but the properties editor: "Issue is that Properties space has its own 'current scene' storage, which is returned from its context, and that gets updated on UI update, not on depsgraph update (as is the case for 'normal' context from other spaces)." according to an administrator in the Blender developers page. So basically it will not update immediately unless there is an UI update.

So for that among some solutions are, maybe the easiest: moving the panel elsewhere like in 3d view, or writing a modal: I tested with a timer modal and it works, but the thing is that a lot of the exporter has to be refactored to avoid race condition.

I implemented the simplest solution and moved the Play and Build project to the 3d View:

![image](https://user-images.githubusercontent.com/42382648/69500755-a742fa80-0edc-11ea-9be4-d624a296879c.png)

I suck at git so I had to create another branch, https://github.com/N8n5h/armory/tree/fix-export
please test the changes to see if that fits as a solution to the original problem.

**Comment by blenderaiser** [comment]
Sorry, I had no time last week.

Nice! Worked well! Particularly I prefer Play and Build in 3d view kkk

... but I noticed that instances of collections are not being exported. Later I update the post with an example.

**Comment by N8n5h** [comment]
Ok nice to hear it works. Regarding that other issue, does that occur without the applied fix too?

**Comment by blenderaiser** [comment]
No, it is happening only in applied fix.

**Comment by N8n5h** [comment]
Ok, I will try to look at it, but seems a bit out of scope of the original issue. The original one was resolved with the fix, right?

**Comment by blenderaiser** [comment]
Yes, about modifiers, I think it was solved, thanks! About the instances of collections, it appears that their objects are not included in the export.

**Comment by N8n5h** [comment]
Ok, please provide an example when you can so I can get testing with it! :+1:

**Comment by blenderaiser** [comment]
Here is the example.
[main.blend.zip](https://github.com/armory3d/armory/files/3889334/main.blend.zip)

The logic below may be able to access objects for export. Hope that helps in something.
``` python
if evaluated_object.instance_type == "COLLECTION":
    for evaluated_object_in_collection in evaluated_object.instance_collection.all_objects:
      

...
```

I would like to use collections as prefabs, because their instances allow editing of transform, as well as preserving the hierarchy of objects (where proxies fall short) and not requiring a link. Is there another way to do this?

**Comment by N8n5h** [comment]
> Is there another way to do this?

I haven't played too much with those aspects of the engine, so I don't know if there is another way. But, what I know is that an asset manager is planned for Blender for not so far future versions, and if I recall correctly it would have some behavior like prefabs similar to Unity, but not sure, I would have to look it up again.

As from the issue, the error seems to be here: https://github.com/N8n5h/armory/blob/e7fb260aeae5f40c5a47cfaa8c9d7b9f98517c1f/blender/arm/exporter.py#L1944

I glossed over it, but it looks like that code is a catch-all for whatever is foreign to the scene, including linked objects, I have include it but make sure linked objects are not picked in the mix...

**Comment by N8n5h** [comment]
Update, I pushed a fix to the fix-export branch https://github.com/N8n5h/armory/commit/674f9bc250a0c013153d618db3013b61191bf60f, now it loops over the collections of each scene instead of all the collections in the blend file, so now it should work fine, I tested a bit to see if everything mentioned in here worked and it seems like everything works. If you find something else please do tell.

**Comment by N8n5h** [comment]
I made another fix with a different approach https://github.com/N8n5h/armory/tree/fix-export-2: 

this time I removed the looping over different scenes and instead I create a temporal collection that will have all the objects and collections from differents scenes linked to it, creating some "zoo" like collection to expose all objects/collections to the dependency graph. 

This let me keep the Play and Build panels over the properties editor. Not sure if this introduces any issues but with the tests from this issue seems to work.
